File: Objective-c/pamservices/PAM.m
//
//  PAM.m
//  pamservices
//
//  Created by heart on 9/8/2560 BE.
//  Copyright © 2560 heart. All rights reserved.
//

#import "PAM.h"
#import "PAMLocalDataBase.h"
#import "PAMClient.h"

@implementation PAM

+(TrackingDataBuilder*)defaultTrackingData{
    static TrackingDataBuilder *defaultData = nil;
    @synchronized(self) {
        if (defaultData == nil)
            defaultData = [[TrackingDataBuilder alloc] init];
    }
    return defaultData;
}

+(TrackingDataBuilder*)createTackingDataBuilder{
    return [[self defaultTrackingData] clone];
}

+(void)trackUpdfh:(NSString*)updfh{
    [[PAMLocalDataBase getInstance] saveUPDFH:updfh];
    [self defaultTrackingData].updfh = updfh;
    [self trackPageViewWithPagename:@"updfh"];
}

+(void)trackPageViewWithPagename:(NSString*)pageName data:(TrackingData*)data{
    data.page_title = pageName;
    [[self defaultPAMClient] trackPageView:data];
}

+(void)trackPageViewWithPagename:(NSString*)pageName{
    TrackingDataBuilder *builder = [self createTackingDataBuilder];
    builder.page_title = pageName;
    [[self defaultPAMClient] trackPageView:[builder build]];
}

+(void)trackPageViewWithData:(TrackingData*)data{
    [[self defaultPAMClient] trackPageView:data];
}

+(void)trackCustomField:(NSDictionary*)customField{
    TrackingDataBuilder *builder = [self createTackingDataBuilder];
    builder.page_title = @"custom field";
    [[self defaultPAMClient] trackCustomField:[builder build] customField:customField];
}

+(void)initPam:(NSString*)pamUrl appId:(NSString*)appId{
    
    TrackingDataBuilder *defaultData = [self defaultTrackingData];
    defaultData.appId = appId;
    defaultData.counter = 0;
    
    NSString *updfh = [[PAMLocalDataBase getInstance] getUPDFH];
    NSString *mtc_id = [[PAMLocalDataBase getInstance] getMtcID];
    NSString *sid = [[PAMLocalDataBase getInstance] getSID];
    
    if(updfh != nil && ![updfh isEqualToString:@""] ){
        defaultData.updfh = updfh;
    }
    
    if(mtc_id != nil && ![mtc_id isEqualToString:@""]){
        defaultData.mtc_id = mtc_id;
    }
    
    if(sid != nil && ![sid isEqualToString:@""]){
        defaultData.sid = sid;
    }
   
    [self defaultPAMClient].pamUrl = pamUrl;
}

+(PAMClient*)defaultPAMClient{
    static PAMClient *pam = nil;
    @synchronized(self) {
        if (pam == nil)
            pam = [[PAMClient alloc] init];
    }
    return pam;
}


@end

File: Swift/pamservices/PAM.swift
//
//  PAM.swift
//  pamservices
//
//  Created by narongrit kanhanoi on 9/8/2560 BE.
//  Copyright © 2017 pushandmotion.com. All rights reserved.
//

public class PAM {
    class func defaultTrackingData() -> TrackingDataBuilder {
        var defaultData: TrackingDataBuilder? = nil
        let lockQueue = DispatchQueue(label: "self")
        lockQueue.sync {
            if defaultData == nil {
                defaultData = TrackingDataBuilder()
            }
        }
        return defaultData!
    }
    
    class func createTackingDataBuilder() -> TrackingDataBuilder {
        return self.defaultTrackingData().clone()
    }
    
    public class func trackUpdfh(_ updfh: String) {
        PAMLocalDataBase.getInstance().saveUPDFH(updfh)
        self.defaultTrackingData().updfh = updfh
        self.trackPageView(withPagename: "updfh")
    }
    
    class public func trackPageView(withPagename pageName: String, data: TrackingData) {
        data.page_title = pageName
        self.defaultPAMClient().trackPageView(data)
    }
    
    public class func trackPageView(withPagename pageName: String) {
        let builder: TrackingDataBuilder? = self.createTackingDataBuilder()
        builder?.page_title = pageName
        self.defaultPAMClient().trackPageView(builder!.build())
    }
    
    public class func trackPageView(with data: TrackingData) {
        self.defaultPAMClient().trackPageView(data)
    }
    
    public class func trackCustomField(cuttomFields:Dictionary<String,String>){
        let builder: TrackingDataBuilder? = self.createTackingDataBuilder()
        builder?.page_title = "cuttomFields"
        self.defaultPAMClient().trackPageView(builder!.build(), customFields: cuttomFields)
    }
    
    public class func initPam(_ pamUrl: String, appId: String) {
        let defaultData: TrackingDataBuilder? = self.defaultTrackingData()
        defaultData?.appId = appId
        defaultData?.counter = 0
        
        let updfh = PAMLocalDataBase.getInstance().getUPDFH()
        let mtc_id = PAMLocalDataBase.getInstance().getMtcID()
        
        let sid: String = PAMLocalDataBase.getInstance().getSID()
        if updfh != "" {
            defaultData?.updfh = updfh
        }
        if mtc_id != "" {
            defaultData?.mtc_id = mtc_id
        }
        if sid != "" {
            defaultData?.sid = sid
        }
        self.defaultPAMClient().pamUrl = pamUrl
    }
    
    public class func defaultPAMClient() -> PAMClient {
        var pam: PAMClient? = nil
        let lockQueue = DispatchQueue(label: "self")
        lockQueue.sync {
            if pam == nil {
                pam = PAMClient()
            }
        }
        return pam!
    }
}
